core/src/main/scala/org/http4s/HeaderKey.scala

```diff
@@ -15,6 +17,8 @@ sealed trait HeaderKey {
   final def unapply(header: Header): Option[HeaderT] = matchHeader(header)
 
   override def toString: String = s"HeaderKey($name})"
+
+  def fromString(s: String): ParseResult[HeaderT]
```

Comment:
I think this could have a default implementation that is the combination of [`HttpHeaderParser.parseHeader`](https://github.com/http4s/http4s/blob/topic/header-laws/core/src/main/scala/org/http4s/parser/HttpHeaderParser.scala#L69) followed by [`matchHeader`](https://github.com/http4s/http4s/blob/topic/header-laws/core/src/main/scala/org/http4s/HeaderKey.scala#L16) to ensure the correct type.

Making explicit overrides does save the indirection of looking through the parser `ConcurrentHashMap` and the follow up check but I don't see much of a use case beyond testing.

Its also not clear from the function name if the input string needs to have the name or not (and it should not based on the implementation).